### PR TITLE
[NFC] Fix GCC 13 warnings

### PR DESCRIPTION
**Pull request by ncesario-lunarg**
Fixes some of the warnings found when compiling on GCC 13.

### CODE REVIEW

File: include/llvm/Analysis/DxilValueCache.h [hunk 1]
```diff
@@ -51,8 +51,7 @@ struct DxilValueCache : public ImmutablePass {
   };
 
 private:
-
-  WeakValueMap ValueMap;
```
Comment (ncesario-lunarg) [inline include/llvm/Analysis/DxilValueCache.h:55]: Hides the `ValueMap` above.

File: tools/clang/lib/Parse/ParseExpr.cpp [hunk 2]
```diff
@@ -446,24 +446,28 @@ Parser::ParseRHSOfBinaryExpression(ExprResult LHS, prec::Level MinPrec) {
...
+    } else {
       // Ensure potential typos in the RHS aren't left undiagnosed.
       Actions.CorrectDelayedTyposInExpr(RHS);
-      // HLSL Change Begin - Take care TernaryMiddle.
-      Actions.CorrectDelayedTyposInExpr(TernaryMiddle);
-      // HLSL Change End.
+    }
+
+    // HLSL Change Begin - Take care TernaryMiddle.
+    Actions.CorrectDelayedTyposInExpr(TernaryMiddle);
```
Comment (ncesario-lunarg) [inline tools/clang/lib/Parse/ParseExpr.cpp:469]: Was this meant to be part of the else clause?

**Comment by llvm-beanz** [comment]
Some of these changes were already proposed in #5179.

File: lib/ProfileData/SampleProfWriter.cpp [hunk 3]
```diff
@@ -122,5 +122,5 @@ SampleProfileWriter::create(StringRef Filename, SampleProfileFormat Format) {
   if (EC)
     return EC;
 
-  return std::move(Writer);
```
Comment (llvm-beanz) [inline lib/ProfileData/SampleProfWriter.cpp:125]: This likely cuts both ways. Clang and GCC disagree about when std::move is valid, and that behavior has changed based on which C++ language version you're targeting.

File: tools/clang/lib/AST/DeclGroup.cpp [hunk 4]
```diff
@@ -30,5 +30,6 @@ DeclGroup* DeclGroup::Create(ASTContext &C, Decl **Decls, unsigned NumDecls) {
 DeclGroup::DeclGroup(unsigned numdecls, Decl** decls) : NumDecls(numdecls) {
   assert(numdecls > 0);
   assert(decls);
-  memcpy(this+1, decls, numdecls * sizeof(*decls));
+  memcpy(reinterpret_cast<void *>(this + 1),
```
Comment (llvm-beanz) [inline tools/clang/lib/AST/DeclGroup.cpp:33]: `static_cast` is preferred when converting to `void*` as it is legal for the narrowest valid cast.

File: tools/clang/lib/Parse/ParseExpr.cpp [hunk 5]
```diff
@@ -446,24 +446,28 @@ Parser::ParseRHSOfBinaryExpression(ExprResult LHS, prec::Level MinPrec) {
         // If we're using '>>' as an operator within a template
         // argument list (in C++98), suggest the addition of
         // parentheses so that the code remains well-formed in C++0x.
-        if (!GreaterThanIsOperator && OpToken.is(tok::greatergreater))
+        if (!GreaterThanIsOperator && OpToken.is(tok::greatergreater)) {
```
Comment (llvm-beanz) [inline tools/clang/lib/Parse/ParseExpr.cpp:449]: This change looks like it changes behavior, and not in a correct way. I assume the GCC warning you're fixing here is something about alignment and conditional nesting... I suspect that warning should be disabled.

File: tools/clang/lib/Parse/ParseExpr.cpp (hunk 2 above)
Comment (llvm-beanz) [reply tools/clang/lib/Parse/ParseExpr.cpp:469]: This line probably does nothing ever and should probably just be removed. Typo correction for the middle expression occurs earlier, and HLSL doesn't have anything special here from C/C++ behaviors.

Locally if I remove this block of HLSL-specific code nothing fails, so it should be fine.

**Review by llvm-beanz — CHANGES_REQUESTED** [review]

File: include/llvm/Analysis/DxilValueCache.h (hunk 1 above)
Comment (ncesario-lunarg) [reply include/llvm/Analysis/DxilValueCache.h:55]: Looks like this was taken care of with a98b242cacfb42c0277c5eabe876fbb47a885f60.

File: lib/ProfileData/SampleProfWriter.cpp (hunk 3 above)
Comment (ncesario-lunarg) [reply lib/ProfileData/SampleProfWriter.cpp:125]: :+1:  FWIW, clang 16.0.3 appears to be "happy" with and without `std::move`.

File: tools/clang/lib/Parse/ParseExpr.cpp (hunk 5 above)
Comment (ncesario-lunarg) [reply tools/clang/lib/Parse/ParseExpr.cpp:449]: Sorry, I'm not seeing how this changes functionality? In any case I will revert as I do not want to incur undue confusion. I found this code rather difficult to read, but that could just be me.

**Comment by ncesario-lunarg** [comment]
Closing this as it looks like a lot of these proposed changes have already merged.

**Pull request closed without merging**